span-builder/service.py

```diff
@@ -0,0 +1,203 @@
+    if val_span_name not in data_store[val_trace_id][val_node_id][val_thread_id]:
+        data_store[val_trace_id][val_node_id][val_thread_id][val_span_name] = {}
+    return data_store[val_trace_id][val_node_id][val_thread_id][val_span_name]
+
+
+def _send_trace_to_jaeger(payload):
+    print("Sending payload to Jaeger: ", payload)
+    requests.post(JAEGER_ENDPOINT, json=payload)
+
+
+def _add_parent_child_relationship(payload):
+    # Parent criteria:
+    # 1. parent's starttime <= child's starttime, and, parent's endtime >= child's endtime
+    # 2. parent's span should be the tightest span that satisfies criteria 1
+
+    spans = payload["resourceSpans"][0]["scopeSpans"][0]["spans"]
+    for span in spans:
+        parent_starttime = float("-inf")
+        parent_endtime = float("inf")
+        for other in spans:
+            if span["spanId"] == other["spanId"]:
+                continue
+            if int(other["startTimeUnixNano"]) <= int(
+                span["startTimeUnixNano"]
+            ) and int(other["endTimeUnixNano"]) >= int(span["endTimeUnixNano"]):
+                if (
+                    int(other["startTimeUnixNano"]) > parent_starttime
+                    and int(other["endTimeUnixNano"]) < parent_endtime
+                ):
+                    span["parentSpanId"] = other["spanId"]
+                    parent_starttime = int(other["startTimeUnixNano"])
+                    parent_endtime = int(other["endTimeUnixNano"])
+    return payload
+
+
+def _is_trace_complete(val_trace_id):
+    # TODO: determine when a trace is complete
+    # criteria: function START should be closed by a function END
+    #
+    # However, here's a counter example:
+    # Let's say we have trace1, comprising of span1 and span2,
+    # which are non-overlapping in time.
+    # We can have children like span1-1, span1-2, whatever.
+    # When the span builder receives only span1-1, span1-2, and span1,
+    # what makes the span builder wait for span2?
+    # Because it can just send trace1->span1->{span1-1, span1-2} and that would be a valid trace
+    # In this case we'll be missing span2 information
+
+    with data_store_locks[val_trace_id]:
+        for val_node_id in data_store[val_trace_id]:
+            for val_thread_id in data_store[val_trace_id][val_node_id]:
+                for val_span_name in data_store[val_trace_id][val_node_id][
+                    val_thread_id
+                ]:
+                    if (
+                        Stage.START.name
+                        not in data_store[val_trace_id][val_node_id][val_thread_id][
+                            val_span_name
+                        ]
+                        or Stage.STOP.name
+                        not in data_store[val_trace_id][val_node_id][val_thread_id][
+                            val_span_name
+                        ]
+                    ):
+                        return False
+    return True
```

Comment:
This is a good first pass for our purposes. I agree with the counter example you posted. I'm checking w/ Sushant to see if we should log when the trace starts and ends so we can capture this concretely instead of inferring.